### ISSUE TITLE
🧹Simplify API and generate vanilla Playwright code 🍦(with a few small helpers)

### ISSUE THREAD

**Issue by jperl** [qawolf/qawolf#432]
Goals

- Generate vanilla Playwright code
- as small API as possible
- clear line between qawolf and playwright. do not decorate or wrap playwright helpers

Generate pure Playwright code, except in corner cases which are not possible with vanilla playwright yet (multiple pages/alternate frames).

Prerequisite: #398 being able to register our selector engine 🙏 

API:

```
const { scroll, launch, $page } = require("qawolf");

// playwright launch + capture and browser choice via env variable
const browser = await launch();

// pure playwright
const page = await browser.newPage();

// $page helper auto-waits for the page (and in the future frame)
// Can pass creation index or page object and it will auto-wait for page
const page = await $page(1 || page);

// missing methods we fill in
await scroll(page, value);
```

Implementation Ideas

- Remove wrapper code by hooking into prototype constructor of context/page. Set up global registries.

**Comment by jperl** [comment]
We could even have a flag that generates pure playwright to make our helpers opt in

**Comment by jperl** [comment]
const { $ } = require(“qawolf”);

$ if awaited returns element handle.
Chained calls qawolf version of methods.

$(selector, { page, frame }).click();

**Comment by jperl** [comment]
Things I don’t like about our current API

- we decorate stuff
- how does it map to playwright

**Comment by jperl** [comment]
We should wait on doing anything until we learn about whether playwright is getting rid of the context concept.

**Comment by jperl** [comment]
[Using extension methods](https://github.com/microsoft/playwright/issues/1072), doesn't seem as clean

```
const page = await context.qpage({ index: 1, timeout: 0 });
const frame = await page.q$("#myselector")
await (page || frame).qclick(selector, options);
```

**Comment by jperl** [comment]
Update scroll to match the same format as Playwright.

**Comment by jperl** [comment]
Consider how to improve type based on conversation with @codex-zaydek.

Could separate "type" and "keys". Keys could have a repeat: n option.

**Comment by jperl** [comment]
Make sure users can use jest-playwright safely

**Comment by codex-zaydek** [comment]
Is the simplified API documented yet? I can provide my input for what I think it should look like. In my mind there’s only 10 or so functions a user should need to learn/know by heart in order to get started. The rest are more less implementation details.

I have a good sense of what this should look like because I’ve had to create a subset myself with Playwright: https://github.com/codex-src/codex-app/blob/9d18903172ae660d077809bb9935a91e13ccdec8/src/components/Editor/__tests/playwright.js#L97 for input events. I also just learned Cypress’s API, which is crazy simple. As simple as learning jQuery for the first time.

**Comment by jperl** [comment]
The API will only be methods QA Wolf generates so:

```
$page.click(selector);
$page.select(selector, value);
$page.scroll(value);
$page.type(selector, value);
```

I would like to make the API as close to Playwright as possible. This way you can remove $ and have vanilla Playwright code.

$page.click defaults to the first page opened. You can use $page(1).click to scope actions for another page.

The only tricky API is type, which needs to be figured out.

**Comment by jperl** [comment]
My thoughts for type are

`$page.type(selector, value, options)`

I like the idea of changing our parsing keys to a whitelist of sequences [like the Cypress API](https://docs.cypress.io/api/commands/type.html#Arguments) you pointed out.

$page.type(`hello{4*backspace}`);

**Comment by codex-zaydek** [comment]
Ultimately I ended up doing something like this because I needed to reuse the grapheme-string length multiple times:

```js
		const data = "😀😃😄😁😆😅🤣😂🙂🙃😉😊😇"
		cy.get("#editor")
			.type(data)
			.getCodex(data)
		cy.get("#editor")
			.backspace(len(data))
			.getCodex("")
```

So in this case, it was preferred to have a separate function for backspace but _still_ have access to a multiplier, with `1` being the default value. 

I think the benefit of string serialization is ease of use, and being able to keep things as concise as one line. Despite that, it’s more work. I do think your inline multiplier idea is really clever.

**Comment by codex-zaydek** [comment]
```
$page.click(selector, options);
$page.select(selector, value, options);
$page.scroll(value, options);
$page.type(selector, value, options);
```

So these are the only methods you want to start with? What’s the deal with `$`? Is it needed?

I like the idea that you can somehow invoke Playwright as an escape hatch, but other than that I see little need to conflate the two. It’s your call, but I’m a bit sick of the Playwright API, the less I have to think _in terms of Playwright unless I have to_, I think the better.

**Comment by jperl** [comment]
My only beef with the Playwright API is their auto-waiting.

$page is a global. The point of the $, is to make it easy to switch between a Playwright page, and the global $page helper which uses the QA Wolf auto-waiting code. For example we will generate this code:

```
const { $page, launch } = require("qawolf");
const selectors = require("../selectors/learn");

describe("learn", () => {
  let page;
  
  beforeAll(async () => {
    const context = await launch({ url: process.env.URL });
    page = await context.newPage();
  });

  afterAll(() => context.close());

  it('can click "Learn React" link', async () => {
    await $page.click(selectors[0]);
  });

  it('can click "Docs" link', async () => {
    await $page.click(selectors[1]);
  });
});
```

By removing $ you have pure playwright code.

**Comment by codex-zaydek** [comment]
Oh, so `$page` refers to QA Wolf, not Playwright. I actually assumed the opposite because `$page` is uglier, feels more raw.

I would do the opposite if it’s not too late (if my previous comment is accurate). Let users spend 99% of their time with `page` and use `$page` as an escape hatch. Maybe that’s confusing because jQuery uses `$` but I still think it’s uglier.

Hope that helps!

**Comment by jperl** [comment]
Good point, we can generate this instead

```
const { $page, launch } = require("qawolf");
const selectors = require("../selectors/learn");

describe("learn", () => {
  let page;
  
  beforeAll(async () => {
    const context = await launch({ url: process.env.URL });
    page = await context.newPage();
    // comment this out for escape hatch
    page = $page;
  });

  afterAll(() => context.close());

  it('can click "Learn React" link', async () => {
    await page.click(selectors[0]);
  });

  it('can click "Docs" link', async () => {
    await page.click(selectors[1]);
  });
});
```

**Comment by jperl** [comment]
I guess it is a tradeoff that might be confusing to users. I want there to be a clear line between qawolf code vs playwright code, since we get that question. That is where the $ comes in.

**Comment by jperl** [comment]
I guess we could do something like this, but I think it is more confusing.

```
const { $page as page, launch } = require("qawolf");
const selectors = require("../selectors/learn");

describe("learn", () => {
  let context;
  let $page;
  
  beforeAll(async () => {
    context = await launch({ url: process.env.URL });
    $page = await context.newPage();
  });

  afterAll(() => context.close());

  it('can click "Learn React" link', async () => {
    await page.click(selectors[0]);
  });
});
```

**Comment by codex-zaydek** [comment]
Yeah, I don’t like having to alias everytime. Here’s what I recommend:

```
const launch = require("qawolf");
const selectors = require("../selectors/learn");

describe("learn", () => {
  let browser;
  let page;
	let $page; // Escape hatch; use as needed, optional
  
  beforeAll(async () => {
    browser = await launch(url);
		// This also works if a user doesn’t need an escape hatch
	  // [page] = await browser.newPage()
    [page, $page] = await browser.newPage()
  });

  afterAll(async () => {
		await browser.close()
	});

  it('can click "Learn React" link', async () => {
    await page.click(selectors[0]);
  });
});
```

Can you use a string literal as a selector as well? I don’t know if I want these exclusively hard-coded in advance. You may consider making the selector feature (where it saves it locally) a configurable option, then tests are always just one file. For me, I would prefer to use `data-qw` literal selectors that won’t change between tests. If I’m a new user, it would also nice to not have to refactor my app, so I think there’s a really good argument for selectors, but perhaps not always.

**Comment by jperl** [comment]
await browser.newPage() is a Playwright API we do not have control over, and $page will not have any other helpers besides those generated. So you will still want access to the Playwright page.

--

Yes, we are switching to strings from objects as part of this API #398. We inline the attribute selectors, and I will make sure we discard the selectors file if there are no html selectors in #376.

We have thought a lot about inlining all selectors. I am pro it for our codebase simplicity 😄 @flaurida is against it because a lot of websites have crazy html and the code file will look really bad.

**Comment by codex-zaydek** [comment]
> await browser.newPage() is a Playwright API we do not have control over, and $page will not have any other helpers besides those generated. So you will still want access to the Playwright page.

If you really wanted to, you could obscure Playwright here and import something other than _exactly_ Playwright’s `launch/context`. As long as you provide an escape hatch, you can do whatever you want. I just think you really want to protect your users from Playwright. It’s a nightmare. 😱 That’s why I’m in favor of doing what you can to protect the user from direct interaction with Playwright.

> We have thought a lot about inlining all selectors. I am pro it for our codebase simplicity 😄 @flaurida is against it because a lot of websites have crazy html and the code file will look really bad.

I think Laura is right because _you are asking users to refactor their sites_ in order to use QA Wolf. I actually think this is the right premise to work off of because you can’t assume your websites _only_ want to test sites they have control over; what if a company hires a contractor to write tests for their website? You need to be flexible enough to support _both_ kinds of users in my opinion.

I think you need to support both here. It’s annoying, yes, but if does serve both ends of the same spectrum.

**Comment by jperl** [comment]
Playwright's API is based off of Puppeteer which has a large community we want to embrace. It has room for improvement (all APIs do) and they are actively working on that.

I want QA Wolf to focus on authoring the tests for you which is a large enough task on its own. So I view our goal with the API as a stop gap where necessary, not a wholesale replacement.

**Comment by jperl** [comment]
Updated ^ with slimmer API. Turns out Playwright page APIs will retry across navigation errors, we were only experiencing them due to using ElementHandle methods.

**Comment by codex-zaydek** [comment]
The reason I bring all of this up is because I can’t use Playwright as is — I need to essentially write a simple utility-library on top of Playwright because the API is too terse and in some cases broken. Cypress on the other hand has a dead simple API. So I don’t know if what I’m thinking about is an open source _extension_ to QA Wolf or something more central, but my hope is that someone else in the future wouldn’t find themselves in the same position as me.

For my editor, I need Playwright because it’s as close to the real thing as possible. Cypress cheats a little with user input which makes it useless to test with my editor.

So Playwright is strictly for testing my editor: https://github.com/codex-src/codex-playwright and Cypress is for everything else: https://github.com/codex-src/codex-cypress.

Because the Playwright API is really bare-metal, I implemented the following in order to actually be _productive_ with Playwright (you get this kind of thing for free with Cypress): https://github.com/codex-src/codex-playwright/blob/master/lib/playwright/playwright.js.

You don’t need to bake these kind of high-level APIs into QA Wolf, _but strictly speaking as a user_ this is what I expect from a e2e testing library. I actually don’t even know if I consider Playwright a e2e testing library because you need to seriously invest in learning Playwright because you can get anything done.

I don’t want to see QA Wolf follow Playwright too closely because then I end up needing to rearchitect all of these helper utilities which are _essentially_ the only thing i care about in order to be productive.

**Comment by jperl** [comment]
Yes we will definitely have helper utilities in QA Wolf to simplify the code generated.

**Comment by codex-zaydek** [comment]
Ah, I may be barking up the wrong issue. 😅

**Comment by jperl** [comment]
Closing this in favor of #457
